src/client/event-listeners.js

```diff
@@ -56,51 +56,72 @@ export default class EventListenerManager {
 
     // If context is specified, addEventListener is called using that as the 'this' object. Otherwise the current this is used.
-    var self = this;
-    var dispatchMouseEventsViaDOM = false;
-    var dispatchKeyEventsViaDOM = false;
+    const self = this;
+    const dispatchMouseEventsViaDOM = false;
+    const dispatchKeyEventsViaDOM = false;
+
     function replaceEventListener(obj, context) {
-      var realAddEventListener = obj.addEventListener;
+      const realAddEventListener = obj.addEventListener;
       obj.addEventListener = function(type, listener, useCapture) {
         self.ensureNoClientHandlers();
-        if (overriddenMessageTypes.indexOf(type) != -1) {
-          var registerListenerToDOM =
+        if (overriddenMessageTypes.indexOf(type) !== -1) {
+          const registerListenerToDOM =
                (type.indexOf('mouse') === -1 || dispatchMouseEventsViaDOM)
             && (type.indexOf('key') === -1 || dispatchKeyEventsViaDOM);
-          var filteredEventListener = function(e) { try { if (e.programmatic || !e.isTrusted) listener(e); } catch(e) {} };
+
+          const filteredEventListener = function(e) {
+            try {
+              if (e.programmatic || !e.isTrusted) listener(e);
+            } catch(e) {}
+          };
           //!!! var filteredEventListener = listener;
-          if (registerListenerToDOM) realAddEventListener.call(context || this, type, filteredEventListener, useCapture);
+
+          if (registerListenerToDOM) {
+            realAddEventListener.call(context || this, type, filteredEventListener, useCapture);
+          }
+
           self.registeredEventListeners.push({
             context: context || this,
             type: type,
             fun: filteredEventListener,
-            useCapture: useCapture
+            realFun: listener,
+            useCapture: useCapture,
+            added: registerListenerToDOM
           });
         } else {
           realAddEventListener.call(context || this, type, listener, useCapture);
           self.registeredEventListeners.push({
             context: context || this,
             type: type,
             fun: listener,
-            useCapture: useCapture
+            realFun: listener,
+            useCapture: useCapture,
+            added: true
           });
         }
       }
 
-      var realRemoveEventListener = obj.removeEventListener;
-
+      const realRemoveEventListener = obj.removeEventListener;
       obj.removeEventListener = function(type, listener, useCapture) {
-        // if (registerListenerToDOM)
-        //realRemoveEventListener.call(context || this, type, filteredEventListener, useCapture);
-        for (var i = 0; i < self.registeredEventListeners.length; i++) {
-          var eventListener = self.registeredEventListeners[i];
-          if (eventListener.context === this && eventListener.type === type && eventListener.fun === listener) {
-            self.registeredEventListeners.splice(i, 1);
```

Comment:
Here has two problems.

1. `listener` is raw listener but `eventListener.fun` can be `filteredEventListener`. `listener` should be compared with raw listener.

2. If event listener is found, it's removed from `registeredEventListeners` by using `.splice()`. But loop var `i` doesn't care so iteration can skip an event listener next to a removed event listener.